corehq/apps/es/transient_util.py

```diff
@@ -0,0 +1,103 @@
+"""Transient utilities needed during the interim of converting HQ Elastic logic
+to use the new "client adapter" models.
+"""
+
+from .client import ElasticDocumentAdapter
+from .registry import get_registry
+
+
+def get_adapter_mapping(adapter):
+    """Temporary function for fetching the Elastic mapping (still defined in
+    pillowtop module) for an adapter.
+    """
+    return _DOC_MAPPINGS_BY_INDEX[(adapter.index_name, adapter.type)]
+
+
+def from_dict_with_possible_id(doc):
+    """Temporary "common" function for adapters who don't yet own their document
+    ``from_python()`` logic.
+    """
+    if "_id" in doc:
+        return doc["_id"], {k: doc[k] for k in doc if k != "_id"}
+    return None, doc
+
+
+def doc_adapter_from_info(index_info, for_export=False):
+    """Return the document adapter for the provided ``index_info`` object.
+
+    :param index_info: instance of a pillowtop ``ElasticsearchIndexInfo`` object
+    :param for_export: ``bool`` used to instantiate the adapter instance
+    :returns: instance of an ``ElasticDocumentAdapter`` subclass
+    """
+    return _DOC_ADAPTERS_BY_INDEX[index_info.index](for_export=for_export)
+
+
+def doc_adapter_from_alias(index_alias, for_export=False):
+    """Return the document adapter for the provided ``index_alias``.
+
+    :param index_alias: ``str`` name of a valid alias assigned to an index
+    :param for_export: ``bool`` used to instantiate the adapter instance
+    :returns: instance of an ``ElasticDocumentAdapter`` subclass
+    """
+    return _DOC_ADAPTERS_BY_ALIAS[index_alias](for_export=for_export)
+
+
+def report_and_fail_on_shard_failures(search_result):
+    ElasticDocumentAdapter._report_and_fail_on_shard_failures(search_result)
+
+
+def _populate_doc_adapter_map(is_test):
+    from .apps import ElasticApp
+    from .case_search import ElasticCaseSearch
+    from .cases import ElasticCase, ElasticReportCase
+    from .domains import ElasticDomain
+    from .forms import ElasticForm, ElasticReportForm
+    from .groups import ElasticGroup
+    from .sms import ElasticSMS
+    from .users import ElasticUser
+
+    # by index
+    for DocAdapter in [ElasticApp, ElasticCaseSearch, ElasticCase,
+                       ElasticReportCase, ElasticDomain, ElasticForm,
+                       ElasticReportForm, ElasticGroup, ElasticSMS,
+                       ElasticUser]:
+        assert DocAdapter.index_name not in _DOC_ADAPTERS_BY_INDEX, \
+            (DocAdapter.index_name, _DOC_ADAPTERS_BY_INDEX)
+        _DOC_ADAPTERS_BY_INDEX[DocAdapter.index_name] = DocAdapter
+    # aliases and mappings
+    for index_info in get_registry().values():
+        _DOC_ADAPTERS_BY_ALIAS[index_info.alias] = _DOC_ADAPTERS_BY_INDEX[index_info.index]
+        mapping_key = (index_info.index, index_info.type)
+        _DOC_MAPPINGS_BY_INDEX[mapping_key] = index_info.mapping
+
+    if is_test:
```

Comment:
> Can this parameter be handled automatically by settings.UNIT_TESTING?

It could, but it's not for Separation of concerns. The importance of that model in this specific scope is limited, but it makes no difference for other reasons so I defaulted to use that design principle.  This function only gets called from [one place](https://github.com/dimagi/commcare-hq/blob/cc8c4d31e85/corehq/apps/es/app_config.py#L11).

The lookups dicts being built here are for very temporary utility.  I think it's advantageous to keep all the logic in one function for clarity and clean removal later.

Reply:
Can you explain a bit more how that principle applies here? The function still needs to change behavior based on whether or not it is in a test, and the standard way to do that in HQ is to checking `settings.UNIT_TESTING`. It is highly unusual to pass that as a parameter to a function in hq, and makes it easier to accidentally misuse the function. It seems especially unnecessary to parameterize this since that parameter is hardcoded to the result of the suggested settings value.

I think the mixing of test and production code is generally a code smell, and as mentioned before would prefer the firmer separation, and confining of test only code to the test modules as per our code conventions. That is both because it makes it easier to find the test code, and because it makes it the production code read more cleanly. However, if you feel strongly that the two setups must be in the same function, I would feel a lot better if the separation was enforced via the settings value rather than a supplied parameter so that it is not possible to accidentally run the test setup in production or otherwise misuse this function.

Reply:
> Can you explain a bit more how that principle applies here?

Writing functions that change their behavior based on global variables is generally something I try to avoid because it makes code more fragile, often harder to grok, and more challenging to test (e.g. requires extra `mock.patch()` cruft).  This is why I implemented the `settings.UNIT_TESTING` value as an argument rather than making the function reach out into the world and fetch it directly (i.e. based on a general coding practice).  This function is only called in one place, and it would be problematic to call this function from somewhere other than the AppConfig's `ready()` method (I cannot imagine a scenario where doing so would be useful), which is why passing the value as an argument was safe. As of d368f5f4bc4 there is a usage note in this function's docstring as well.

Per request, I've changed this to follow existing conventions in ff7351ded65.